Topic/virt x86 segment

liujing2 opened the pull request:
@mcastelino I updated some today which rebased the little fix of your patch and added two more ones.
The current status is,
1.	Specify the bus of each virtio blk and serial devices to 0.pcie.0, but early output shows kernel didn’t add the devices to root bus, because it reads config failed. Key function in kernel is pci_scan_device()->pci_bus_read_dev_vendor_id(), getting vendor-id =ffffffff. I haven’t got idea about this.
2.	I am thinking and trying some logic to separate the pci-hole64 for each segment. 
3.	I need add more patches to disable IO spaces allocations for segment >= 1, and to disable IRQ in pci-lite.c for segment >= 1.
4.	Fw_cfg needs add a new macro to transfer PCI_HOST_PCIEXBAR_BASE. I will look at that later.

I have no time to debug more today, and I'll continue tomorrow!

Commit:
Add "segment_nr" property for VirtMachineState

Users can specify the number of pci host bridge when launching
guest by "-machine virt,segment-nr=N" property.

Signed-off-by: Jing Liu <jing2.liu@linux.intel.com>

Commit:
BUG fix for upstream: Fix checking when setting property

Qemu uses a static bool to avoid the property being set twice but it doesn't
check the owner. This might be because originally there is only one PCI Host
Bridge. For different PCI Host Bridge, ACPI_PCIHP_PROP_BSEL should be
added to each root bus.

Signed-off-by: Jing Liu <jing2.liu@linux.intel.com>

Commit:
MCFG: Make MCFG build more generic

ACPI builds MCFG for single PCI host before because QEMU assumes only
one PCI host on pc and q35 platform. Making MCFG build more generic
by giving the PCI host list. Then MCFG table can include multiple
allocations to reside multiple PCI host information.

NOTE: One case needs discuss:
If one of the PCI hosts doesn't have MCFG_BASE address, just return a
NULL MCFG table, does this make sense?

Signed-off-by: Jing Liu <jing2.liu@linux.intel.com>

Commit:
Set PC and q35 segment as one

On PC and q35 platform, we set it to one.

Signed-off-by: Jing Liu <jing2.liu@linux.intel.com>

Commit:
pci-lite: Make PCI host initialization for multiple host bridge

Once we want to create multiple PCI host for virt platform, we need
use pci_lite_init to help create multiple PCI host and root bus.
Use 0.pci.0 as the name for the first PCI host.

Don't add IO Ports and IRQs from the secondary segment

Later, we will set ACPI related functions as list.

Signed-off-by: Jing Liu <jing2.liu@linux.intel.com>

Commit:
[WIP]pcihp: Make ACPI PCIHP for multiple root bus

Set different names for ACPI PCIHP memory region
This doesn't work now.

Signed-off-by: Jing Liu <jing2.liu@linux.intel.com>

Commit:
Add property check before set

Qemu assumes only one PCI host so it doesn't check the property
adding of ACPI_PCIHP_IO_BASE_PROP. For virt platform, when multiple
PCI segment is used and ACPI PCI hotplug is enabled, it needs check
if this property is already set before.

Signed-off-by: Jing Liu <jing2.liu@linux.intel.com>

Commit:
Build ACPI tables for multiple root bus

Make PCI host as a list when build hardware reduced ACPI for virt platform,
so that it builds DSDT and MCFG referring to the amount of PCI segment.

Signed-off-by: Jing Liu <jing2.liu@linux.intel.com>

Commit:
PCIEXBAR: Removing hardcoding address for each PCI host

Reduce the low memory size to 1G so that we reserve 3G for PCI.
For multiple PCI segment, the first MCFG base is from 0x40000000
and dynamically allocates the MCFG space for each segment.
The PCI hole is just behind all the MCFG space according to the
segment number.

This corresponds to the same logic in OVMF changes.

Note:
1. We can't enable MCFG above 4G, so only use below 4G
2. If each MCFG use 256M, we could only set 4 PCI segment.

Signed-off-by: Jing Liu <jing2.liu@linux.intel.com>

Commit:
Allocate pci_hole64 for each segment

Set the pci_hole64 start after the previous segment pci_hole64 end address.

Signed-off-by: Jing Liu <jing2.liu@linux.intel.com>

Commit:
Pass PCIEXBAR_BASE and PCI_SEGMENT by fw_cfg

This passes PCIEXBAR_BASE and PCI_SEGMENT via fw_cfg so that
the firware does not hardcode this value for MCFG and PCI hole base address.

Signed-off-by: Jing Liu <jing2.liu@linux.intel.com>

liujing2 commented:
Hi @mcastelino 
Now the multiple segment almost works except some issues which need debug later.
1. If not specified, devices will by default cold plugged onto 1.pcie.0
2. Hotplug is not finished.
3. Discuss the pci_hole64_base definition: whether it should be re-defined by firmware? The detailed can be seen in the [patch](https://github.com/mcastelino/nemu/pull/7/commits/494ec0b7042b1a2c31d18299fad10308023a99b7)

There are also some OVMF changes to allocate pci-hole/pci-hole64 dynamically. 
I will debug 1 and continue 2 then. Look forward to your reply on 3.

Thanks!
Jing

mcastelino commented on hw/i386/virt/virt.c:
```diff
@@ -226,6 +226,8 @@ static void virt_machine_state_init(MachineState *machine)
             
     fw_cfg = fw_cfg_init(machine, smp_cpus, mc->possible_cpu_arch_ids(machine), vms->apic_id_limit);
     fw_cfg_add_i16(fw_cfg, FW_CFG_MACHINE_ID, X86_VIRT);
+    fw_cfg_add_i64(fw_cfg, FW_CFG_PCIEXBAR_BASE, PCI_HOST_PCIEXBAR_BASE);
```
@liujing2 this is one way to do it. Another way, which may be better would be to actually create a file. That will allow us to keep this logic generic and add more fields in the future, one for each segment.

Commit:
pci-host: Get pci_host dynamically on virt platform

Originally, Qemu gets pci_host statically by the pci host name,
while virt platform supports multiple segment so this tries
get pci host dynamically.

Signed-off-by: Jing Liu <jing2.liu@linux.intel.com>
